stan/math/rev/fun/tcrossprod.hpp

```diff
@@ -17,45 +17,27 @@ namespace math {
  * Returns the result of post-multiplying a matrix by its
  * own transpose.
  *
- * @tparam T Type of the matrix (must be derived from \c Eigen::MatrixBase)
+ * @tparam T Type of the matrix
  * @param M Matrix to multiply.
  * @return M times its transpose.
  */
-template <typename T, require_eigen_vt<is_var, T>* = nullptr>
-inline Eigen::Matrix<var, T::RowsAtCompileTime, T::RowsAtCompileTime>
-tcrossprod(const T& M) {
-  if (M.rows() == 0) {
-    return {};
+template <typename T, require_rev_matrix_t<T>* = nullptr>
+inline promote_var_matrix_t<Eigen::MatrixXd, T> tcrossprod(const T& M) {
```

Comment:
@t4c1, @SteveBronder I think there's a problem with either `plain_type_t<T>` or `promote_var_matrix_t` or my use thereof.

If the return type here is `plain_type_t<T>`, you'll get errors compiling `crossprod` ([here](https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-2106/11/pipeline/)). `crossprod(x)` is implemented as `tcrossprod(x.transpose())`.

When I changed the return type to `promote_var_matrix_t<T>`, things worked.

I added pretty_prints for the types:

```
T:  stan::math::test::type_name() [Arg = Eigen::Transpose<const Eigen::Matrix<stan::math::var_value<double>, -1, -1, 0, -1, -1> >
plain_type_t :  stan::math::test::type_name() [Arg = Eigen::Matrix<stan::math::var_value<double>, -1, -1, 1, -1, -1>
promote_var_matrix_t :  stan::math::test::type_name() [Arg = Eigen::Matrix<stan::math::var_value<double>, -1, -1, 0, -1, -1>
```

Is it me or it is a template?

Reply:
Sorry I'll look into this tmrw morning. The jenkins link sends me to

```
./test/unit/math/serializer.hpp:199:11: error: no matching member function for call to 'push_back'

    vals_.push_back(x);

    ~~~~~~^~~~~~~~~

./test/unit/math/serializer.hpp:291:5: note: in instantiation of function template specialization 'stan::test::serializer<stan::math::var_value<double> >::write<Eigen::Matrix<stan::math::var_value<double>, -1, -1, 1, -1, -1> >' requested here

  s.write(x);
```

I'm not seeing how it's related to the return type. Is there a reason you can't use `auto` as the return type?

Reply:
> Is there a reason you can't use auto as the return type

We need explicit return types here because the return type gets build internally as an `arena_t<T_Return>` and you gotta set the return type to `T_Return` to get it to cast to a non-arena thing.

Reply:
Looking at the different template arguments between the types:
```
Eigen::Matrix<stan::math::var_value<double>, -1, -1, 1, -1, -1>
Eigen::Matrix<stan::math::var_value<double>, -1, -1, 0, -1, -1>
```

That third integer is the 'Options' flag, which controls storage order. I would guess (emphasis on the guess) that the ```.transpose()``` operator maps the matrix to row-major storage, so the plain type of the input (and the eventual return type) is a row-major matrix and then compilation fails when you try to return a column-major matrix instead. No idea why ```plain_type_t``` and ```promote_var_matrix_t``` would handle that differently though

Reply:
From the type printouts it seems plain_type_t is returning a row major matrix (yes, it can do that in some cases!). Any function not generalized for expresions is still expecting its inputs to be col major, so I guess that causes compilation failures. 

This means that we shouldn't use plain_type_t (or .eval() and auto) for return values in the few functions that eval expresions, for which Eigen prefer row major types.

Reply:
> No idea why plain_type_t and promote_var_matrix_t would handle that differently though

`plain_type_t` actually gets the plain type for the expression. `promote_var_matrix_t` (and `promote_scalar_t`) just determine a (col_major) matrix with appropriate scalar type.

Reply:
I'm going to take away that we should use `promote_var_matrix_t` for the return type and avoid `plain_type_t` to make sure we aren't getting matrices in unexpected storage orders.

Considering this is a bug-like thing that may have shown up in the wild somewhere, is this something we should test for somehow?

Or is this failure to compile actually just a problem with the test framework that should be fixed?

Reply:
I would consider it a bug, but testing for it is not high priority, as it will cease to be a bug as soon as all functions accept general expressions.

Reply:
Won't it be more of an issue with expressions since they'll propagate the row-major matrix which could lead to unexpected behaviour further down the chain?

Reply:
Functions handling expressions also handle row major matrices, so it won't matter.

Reply:
I was mostly thinking that any looping is assuming column-major ordering (i.e., loops over column first), so some programs could end up slower if a function returned a row-major matrix at some point. But I don't know how much looping we still have at the moment

Reply:
> We need explicit return types here because the return type gets build internally as an arena_t<T_Return> and you gotta set the return type to T_Return to get it to cast to a non-arena thing.

`promote_var_matrix_t` returns a `var<matrix>` or `matrix<var>` so I usually just wrap the return like `return ret_type(ret)`

Reply:
I think we should consider any remaining loops over matrices as potential performance issues and replace them with Eigen expressions.